<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">import luigi
from luigi.contrib.spark import PySparkTask
from typing import List, Dict
from pyspark.sql.functions import collect_set, struct, lit, col, concat, flatten

from impc_etl.jobs.extract.gene_production_status_extractor import (
    GeneProductionStatusExtractor,
)
from impc_etl.workflow.config import ImpcConfig
from pyspark.sql import SparkSession, DataFrame
from impc_etl.jobs.load.solr.gene_mapper import get_gene_core_df, IMITS_GENE_COLUMNS
from impc_etl.workflow.extraction import (
    GeneExtractor,
    AlleleExtractor,
    MGIHomoloGeneExtractor,
    MGIMrkListExtractor,
    OntologyMetadataExtractor,
    ProductExtractor,
)
from impc_etl.workflow.load import (
    ObservationsMapper,
    StatsResultsCoreLoader,
    GenotypePhenotypeCoreLoader,
    ImpcImagesCoreLoader,
)

EXCLUDE_PRODUCT_COLUMNS = [
    &#34;marker_mgi_accession_id&#34;,
    &#34;marker_type&#34;,
    &#34;marker_name&#34;,
    &#34;marker_synonym&#34;,
    &#34;allele_mgi_accession_id&#34;,
    &#34;allele_symbol&#34;,
    &#34;allele_has_issue&#34;,
    &#34;allele_synonym&#34;,
    &#34;associated_product_colony_name&#34;,
    &#34;associated_products_vector_names&#34;,
    &#34;loa_assays&#34;,
]


class ImpcGeneBundleMapper(PySparkTask):
    name = &#34;IMPC_Bundle_Mapper&#34;
    embryo_data_json_path = luigi.Parameter()
    mongodb_database = luigi.Parameter()
    output_path = luigi.Parameter()
    packages = &#34;org.mongodb.spark:mongo-spark-connector_2.12:3.0.1&#34;
    mongodb_connection_uri = luigi.Parameter()
    mongodb_genes_collection = luigi.Parameter()
    mongodb_replica_set = luigi.Parameter()

    @property
    def packages(self):
        return super().packages + super(PySparkTask, self).packages

    def requires(self):
        return [
            GeneExtractor(),
            AlleleExtractor(),
            MGIHomoloGeneExtractor(),
            MGIMrkListExtractor(),
            ObservationsMapper(),
            StatsResultsCoreLoader(raw_data_in_output=&#34;bundled&#34;),
            OntologyMetadataExtractor(),
            GeneProductionStatusExtractor(),
            GenotypePhenotypeCoreLoader(),
            ImpcImagesCoreLoader(),
            ProductExtractor(),
        ]

    def output(self):
        return ImpcConfig().get_target(f&#34;{self.output_path}gene_bundle_parquet&#34;)

    def app_options(self):
        return [
            self.input()[0].path,
            self.input()[1].path,
            self.input()[2].path,
            self.input()[3].path,
            self.embryo_data_json_path,
            self.input()[4].path,
            self.input()[5].path,
            self.input()[6].path,
            self.input()[7].path,
            self.input()[8].path,
            self.input()[9].path,
            self.input()[10].path,
            self.output().path,
        ]

    def main(self, sc, *argv):
        imits_gene_parquet_path = argv[0]
        imits_allele_parquet_path = argv[1]
        mgi_homologene_report_parquet_path = argv[2]
        mgi_mrk_list_report_parquet_path = argv[3]
        embryo_data_json_path = argv[4]
        observations_parquet_path = argv[5]
        stats_results_parquet_path = argv[6]
        ontology_metadata_parquet_path = argv[7]
        gene_production_status_path = argv[8]
        genotype_phenotype_parquet_path = argv[9]
        impc_images_parquet_path = argv[10]
        product_parquet_path = argv[11]
        output_path = argv[12]
        spark = SparkSession(sc)

        imits_gene_df = spark.read.parquet(imits_gene_parquet_path).select(
            IMITS_GENE_COLUMNS
        )
        imits_allele_df = spark.read.parquet(imits_allele_parquet_path)
        mgi_homologene_df = spark.read.parquet(mgi_homologene_report_parquet_path)
        mgi_mrk_list_df = spark.read.parquet(mgi_mrk_list_report_parquet_path)
        mgi_mrk_list_df = mgi_mrk_list_df.select(
            [
                &#34;mgi_accession_id&#34;,
                &#34;chr&#34;,
                &#34;genome_coordinate_start&#34;,
                &#34;genome_coordinate_end&#34;,
                &#34;strand&#34;,
            ]
        )
        embryo_data_df = spark.read.json(embryo_data_json_path, mode=&#34;FAILFAST&#34;)
        observations_df = spark.read.parquet(observations_parquet_path)
        stats_results_df = spark.read.parquet(stats_results_parquet_path)
        ontology_metadata_df = spark.read.parquet(ontology_metadata_parquet_path)
        gene_production_status_df = spark.read.parquet(gene_production_status_path)
        genotype_phenotype_df = spark.read.parquet(genotype_phenotype_parquet_path)
        impc_images_df = spark.read.parquet(impc_images_parquet_path)
        product_df = spark.read.parquet(product_parquet_path)
        gene_df: DataFrame = get_gene_core_df(
            imits_gene_df,
            imits_allele_df,
            mgi_homologene_df,
            mgi_mrk_list_df,
            embryo_data_df,
            observations_df,
            stats_results_df,
            ontology_metadata_df,
            gene_production_status_df,
            False,
        )
        gene_df = gene_df.withColumnRenamed(
            &#34;datasets_raw_data&#34;, &#34;gene_statistical_results&#34;
        )
        gene_df = gene_df.withColumn(
            &#34;_class&#34;, lit(&#34;org.mousephenotype.api.models.GeneBundle&#34;)
        )
        genotype_phenotype_df = genotype_phenotype_df.withColumnRenamed(
            &#34;marker_accession_id&#34;, &#34;mgi_accession_id&#34;
        )
        gp_by_gene_df = genotype_phenotype_df.groupBy(&#34;mgi_accession_id&#34;).agg(
            collect_set(
                struct(
                    *[
                        col_name
                        for col_name in genotype_phenotype_df.columns
                        if col_name != &#34;mgi_accession_id&#34;
                    ]
                )
            ).alias(&#34;gene_phenotype_associations&#34;)
        )
        gene_df = gene_df.join(gp_by_gene_df, &#34;mgi_accession_id&#34;, &#34;left_outer&#34;)

        impc_images_df = impc_images_df.withColumnRenamed(
            &#34;gene_accession_id&#34;, &#34;mgi_accession_id&#34;
        )

        images_by_gene_df = impc_images_df.groupBy(&#34;mgi_accession_id&#34;).agg(
            collect_set(
                struct(
                    *[
                        col_name
                        for col_name in impc_images_df.columns
                        if col_name != &#34;mgi_accession_id&#34;
                    ]
                )
            ).alias(&#34;gene_images&#34;)
        )
        gene_df = gene_df.join(images_by_gene_df, &#34;mgi_accession_id&#34;, &#34;left_outer&#34;)

        products_by_gene = product_df.groupBy(&#34;mgi_accession_id&#34;).agg(
            collect_set(
                struct(
                    *[
                        col_name
                        for col_name in product_df.columns
                        if col_name
                        not in [&#34;mgi_accession_id&#34;] + EXCLUDE_PRODUCT_COLUMNS
                    ]
                )
            ).alias(&#34;gene_products&#34;)
        )

        gene_df = gene_df.join(products_by_gene, &#34;mgi_accession_id&#34;, &#34;left_outer&#34;)
        gene_df = gene_df.withColumn(&#34;_id&#34;, col(&#34;mgi_accession_id&#34;))
        gene_df.write.format(&#34;mongo&#34;).mode(&#34;append&#34;).option(
            &#34;spark.mongodb.output.uri&#34;,
            f&#34;{self.mongodb_connection_uri}/admin?replicaSet={self.mongodb_replica_set}&#34;,
        ).option(&#34;database&#34;, str(self.mongodb_database)).option(
            &#34;collection&#34;, str(self.mongodb_genes_collection)
        ).save()

        ## Create gene search index
        gene_search_df = gene_df.select(
            col(&#34;mgi_accession_id&#34;).alias(&#34;_id&#34;),
            &#34;mgi_accession_id&#34;,
            &#34;marker_name&#34;,
            &#34;human_gene_symbol&#34;,
            &#34;marker_synonym&#34;,
            &#34;assignment_status&#34;,
            &#34;crispr_allele_production_status&#34;,
            &#34;es_cell_production_status&#34;,
            &#34;mouse_production_status&#34;,
            &#34;phenotype_status&#34;,
            &#34;phenotyping_data_available&#34;,
            col(&#34;significant_top_level_mp_terms&#34;).alias(&#34;significant_phenotype_system&#34;),
            col(&#34;not_significant_top_level_mp_terms&#34;).alias(
                &#34;non_significant_phenotype_system&#34;
            ),
            concat(
                &#34;gene_phenotype_associations.mp_term_id&#34;,
                flatten(&#34;gene_phenotype_associations.intermediate_mp_term_id&#34;),
                flatten(&#34;gene_phenotype_associations.top_level_mp_term_id&#34;),
            ).alias(&#34;significant_mp_term_ids&#34;),
            concat(
                &#34;gene_phenotype_associations.mp_term_name&#34;,
                flatten(&#34;gene_phenotype_associations.intermediate_mp_term_name&#34;),
                flatten(&#34;gene_phenotype_associations.top_level_mp_term_name&#34;),
            ).alias(&#34;significant_mp_term_names&#34;),
        )
        gene_search_df = gene_search_df.withColumn(
            &#34;_class&#34;, lit(&#34;org.mousephenotype.api.models.Gene&#34;)
        )
        gene_search_df.write.format(&#34;mongo&#34;).mode(&#34;append&#34;).option(
            &#34;spark.mongodb.output.uri&#34;,
            f&#34;{self.mongodb_connection_uri}/admin?replicaSet={self.mongodb_replica_set}&#34;,
        ).option(&#34;database&#34;, str(self.mongodb_database)).option(
            &#34;collection&#34;, str(self.mongodb_genes_collection) + &#34;_search&#34;
        ).save()
        gene_df.write.parquet(output_path)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper"><code class="flex name class">
<span>class <span class="ident">ImpcGeneBundleMapper</span></span>
<span>(</span><span>*args, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Template task for running an inline PySpark job</p>
<p>Simply implement the <code>main</code> method in your subclass</p>
<p>You can optionally define package names to be distributed to the cluster
with <code>py_packages</code> (uses luigi's global py-packages configuration by default)</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class ImpcGeneBundleMapper(PySparkTask):
    name = &#34;IMPC_Bundle_Mapper&#34;
    embryo_data_json_path = luigi.Parameter()
    mongodb_database = luigi.Parameter()
    output_path = luigi.Parameter()
    packages = &#34;org.mongodb.spark:mongo-spark-connector_2.12:3.0.1&#34;
    mongodb_connection_uri = luigi.Parameter()
    mongodb_genes_collection = luigi.Parameter()
    mongodb_replica_set = luigi.Parameter()

    @property
    def packages(self):
        return super().packages + super(PySparkTask, self).packages

    def requires(self):
        return [
            GeneExtractor(),
            AlleleExtractor(),
            MGIHomoloGeneExtractor(),
            MGIMrkListExtractor(),
            ObservationsMapper(),
            StatsResultsCoreLoader(raw_data_in_output=&#34;bundled&#34;),
            OntologyMetadataExtractor(),
            GeneProductionStatusExtractor(),
            GenotypePhenotypeCoreLoader(),
            ImpcImagesCoreLoader(),
            ProductExtractor(),
        ]

    def output(self):
        return ImpcConfig().get_target(f&#34;{self.output_path}gene_bundle_parquet&#34;)

    def app_options(self):
        return [
            self.input()[0].path,
            self.input()[1].path,
            self.input()[2].path,
            self.input()[3].path,
            self.embryo_data_json_path,
            self.input()[4].path,
            self.input()[5].path,
            self.input()[6].path,
            self.input()[7].path,
            self.input()[8].path,
            self.input()[9].path,
            self.input()[10].path,
            self.output().path,
        ]

    def main(self, sc, *argv):
        imits_gene_parquet_path = argv[0]
        imits_allele_parquet_path = argv[1]
        mgi_homologene_report_parquet_path = argv[2]
        mgi_mrk_list_report_parquet_path = argv[3]
        embryo_data_json_path = argv[4]
        observations_parquet_path = argv[5]
        stats_results_parquet_path = argv[6]
        ontology_metadata_parquet_path = argv[7]
        gene_production_status_path = argv[8]
        genotype_phenotype_parquet_path = argv[9]
        impc_images_parquet_path = argv[10]
        product_parquet_path = argv[11]
        output_path = argv[12]
        spark = SparkSession(sc)

        imits_gene_df = spark.read.parquet(imits_gene_parquet_path).select(
            IMITS_GENE_COLUMNS
        )
        imits_allele_df = spark.read.parquet(imits_allele_parquet_path)
        mgi_homologene_df = spark.read.parquet(mgi_homologene_report_parquet_path)
        mgi_mrk_list_df = spark.read.parquet(mgi_mrk_list_report_parquet_path)
        mgi_mrk_list_df = mgi_mrk_list_df.select(
            [
                &#34;mgi_accession_id&#34;,
                &#34;chr&#34;,
                &#34;genome_coordinate_start&#34;,
                &#34;genome_coordinate_end&#34;,
                &#34;strand&#34;,
            ]
        )
        embryo_data_df = spark.read.json(embryo_data_json_path, mode=&#34;FAILFAST&#34;)
        observations_df = spark.read.parquet(observations_parquet_path)
        stats_results_df = spark.read.parquet(stats_results_parquet_path)
        ontology_metadata_df = spark.read.parquet(ontology_metadata_parquet_path)
        gene_production_status_df = spark.read.parquet(gene_production_status_path)
        genotype_phenotype_df = spark.read.parquet(genotype_phenotype_parquet_path)
        impc_images_df = spark.read.parquet(impc_images_parquet_path)
        product_df = spark.read.parquet(product_parquet_path)
        gene_df: DataFrame = get_gene_core_df(
            imits_gene_df,
            imits_allele_df,
            mgi_homologene_df,
            mgi_mrk_list_df,
            embryo_data_df,
            observations_df,
            stats_results_df,
            ontology_metadata_df,
            gene_production_status_df,
            False,
        )
        gene_df = gene_df.withColumnRenamed(
            &#34;datasets_raw_data&#34;, &#34;gene_statistical_results&#34;
        )
        gene_df = gene_df.withColumn(
            &#34;_class&#34;, lit(&#34;org.mousephenotype.api.models.GeneBundle&#34;)
        )
        genotype_phenotype_df = genotype_phenotype_df.withColumnRenamed(
            &#34;marker_accession_id&#34;, &#34;mgi_accession_id&#34;
        )
        gp_by_gene_df = genotype_phenotype_df.groupBy(&#34;mgi_accession_id&#34;).agg(
            collect_set(
                struct(
                    *[
                        col_name
                        for col_name in genotype_phenotype_df.columns
                        if col_name != &#34;mgi_accession_id&#34;
                    ]
                )
            ).alias(&#34;gene_phenotype_associations&#34;)
        )
        gene_df = gene_df.join(gp_by_gene_df, &#34;mgi_accession_id&#34;, &#34;left_outer&#34;)

        impc_images_df = impc_images_df.withColumnRenamed(
            &#34;gene_accession_id&#34;, &#34;mgi_accession_id&#34;
        )

        images_by_gene_df = impc_images_df.groupBy(&#34;mgi_accession_id&#34;).agg(
            collect_set(
                struct(
                    *[
                        col_name
                        for col_name in impc_images_df.columns
                        if col_name != &#34;mgi_accession_id&#34;
                    ]
                )
            ).alias(&#34;gene_images&#34;)
        )
        gene_df = gene_df.join(images_by_gene_df, &#34;mgi_accession_id&#34;, &#34;left_outer&#34;)

        products_by_gene = product_df.groupBy(&#34;mgi_accession_id&#34;).agg(
            collect_set(
                struct(
                    *[
                        col_name
                        for col_name in product_df.columns
                        if col_name
                        not in [&#34;mgi_accession_id&#34;] + EXCLUDE_PRODUCT_COLUMNS
                    ]
                )
            ).alias(&#34;gene_products&#34;)
        )

        gene_df = gene_df.join(products_by_gene, &#34;mgi_accession_id&#34;, &#34;left_outer&#34;)
        gene_df = gene_df.withColumn(&#34;_id&#34;, col(&#34;mgi_accession_id&#34;))
        gene_df.write.format(&#34;mongo&#34;).mode(&#34;append&#34;).option(
            &#34;spark.mongodb.output.uri&#34;,
            f&#34;{self.mongodb_connection_uri}/admin?replicaSet={self.mongodb_replica_set}&#34;,
        ).option(&#34;database&#34;, str(self.mongodb_database)).option(
            &#34;collection&#34;, str(self.mongodb_genes_collection)
        ).save()

        ## Create gene search index
        gene_search_df = gene_df.select(
            col(&#34;mgi_accession_id&#34;).alias(&#34;_id&#34;),
            &#34;mgi_accession_id&#34;,
            &#34;marker_name&#34;,
            &#34;human_gene_symbol&#34;,
            &#34;marker_synonym&#34;,
            &#34;assignment_status&#34;,
            &#34;crispr_allele_production_status&#34;,
            &#34;es_cell_production_status&#34;,
            &#34;mouse_production_status&#34;,
            &#34;phenotype_status&#34;,
            &#34;phenotyping_data_available&#34;,
            col(&#34;significant_top_level_mp_terms&#34;).alias(&#34;significant_phenotype_system&#34;),
            col(&#34;not_significant_top_level_mp_terms&#34;).alias(
                &#34;non_significant_phenotype_system&#34;
            ),
            concat(
                &#34;gene_phenotype_associations.mp_term_id&#34;,
                flatten(&#34;gene_phenotype_associations.intermediate_mp_term_id&#34;),
                flatten(&#34;gene_phenotype_associations.top_level_mp_term_id&#34;),
            ).alias(&#34;significant_mp_term_ids&#34;),
            concat(
                &#34;gene_phenotype_associations.mp_term_name&#34;,
                flatten(&#34;gene_phenotype_associations.intermediate_mp_term_name&#34;),
                flatten(&#34;gene_phenotype_associations.top_level_mp_term_name&#34;),
            ).alias(&#34;significant_mp_term_names&#34;),
        )
        gene_search_df = gene_search_df.withColumn(
            &#34;_class&#34;, lit(&#34;org.mousephenotype.api.models.Gene&#34;)
        )
        gene_search_df.write.format(&#34;mongo&#34;).mode(&#34;append&#34;).option(
            &#34;spark.mongodb.output.uri&#34;,
            f&#34;{self.mongodb_connection_uri}/admin?replicaSet={self.mongodb_replica_set}&#34;,
        ).option(&#34;database&#34;, str(self.mongodb_database)).option(
            &#34;collection&#34;, str(self.mongodb_genes_collection) + &#34;_search&#34;
        ).save()
        gene_df.write.parquet(output_path)</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li>luigi.contrib.spark.PySparkTask</li>
<li>luigi.contrib.spark.SparkSubmitTask</li>
<li>luigi.contrib.external_program.ExternalProgramTask</li>
<li>luigi.task.Task</li>
</ul>
<h3>Class variables</h3>
<dl>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.embryo_data_json_path"><code class="name">var <span class="ident">embryo_data_json_path</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_connection_uri"><code class="name">var <span class="ident">mongodb_connection_uri</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_database"><code class="name">var <span class="ident">mongodb_database</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_genes_collection"><code class="name">var <span class="ident">mongodb_genes_collection</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_replica_set"><code class="name">var <span class="ident">mongodb_replica_set</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.name"><code class="name">var <span class="ident">name</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.output_path"><code class="name">var <span class="ident">output_path</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
<h3>Instance variables</h3>
<dl>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.packages"><code class="name">var <span class="ident">packages</span></code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@property
def packages(self):
    return super().packages + super(PySparkTask, self).packages</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.app_options"><code class="name flex">
<span>def <span class="ident">app_options</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>Subclass this method to map your task parameters to the app's arguments</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def app_options(self):
    return [
        self.input()[0].path,
        self.input()[1].path,
        self.input()[2].path,
        self.input()[3].path,
        self.embryo_data_json_path,
        self.input()[4].path,
        self.input()[5].path,
        self.input()[6].path,
        self.input()[7].path,
        self.input()[8].path,
        self.input()[9].path,
        self.input()[10].path,
        self.output().path,
    ]</code></pre>
</details>
</dd>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.main"><code class="name flex">
<span>def <span class="ident">main</span></span>(<span>self, sc, *argv)</span>
</code></dt>
<dd>
<div class="desc"><p>Called by the pyspark_runner with a SparkContext and any arguments returned by <code>app_options()</code></p>
<p>:param sc: SparkContext
:param args: arguments list</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def main(self, sc, *argv):
    imits_gene_parquet_path = argv[0]
    imits_allele_parquet_path = argv[1]
    mgi_homologene_report_parquet_path = argv[2]
    mgi_mrk_list_report_parquet_path = argv[3]
    embryo_data_json_path = argv[4]
    observations_parquet_path = argv[5]
    stats_results_parquet_path = argv[6]
    ontology_metadata_parquet_path = argv[7]
    gene_production_status_path = argv[8]
    genotype_phenotype_parquet_path = argv[9]
    impc_images_parquet_path = argv[10]
    product_parquet_path = argv[11]
    output_path = argv[12]
    spark = SparkSession(sc)

    imits_gene_df = spark.read.parquet(imits_gene_parquet_path).select(
        IMITS_GENE_COLUMNS
    )
    imits_allele_df = spark.read.parquet(imits_allele_parquet_path)
    mgi_homologene_df = spark.read.parquet(mgi_homologene_report_parquet_path)
    mgi_mrk_list_df = spark.read.parquet(mgi_mrk_list_report_parquet_path)
    mgi_mrk_list_df = mgi_mrk_list_df.select(
        [
            &#34;mgi_accession_id&#34;,
            &#34;chr&#34;,
            &#34;genome_coordinate_start&#34;,
            &#34;genome_coordinate_end&#34;,
            &#34;strand&#34;,
        ]
    )
    embryo_data_df = spark.read.json(embryo_data_json_path, mode=&#34;FAILFAST&#34;)
    observations_df = spark.read.parquet(observations_parquet_path)
    stats_results_df = spark.read.parquet(stats_results_parquet_path)
    ontology_metadata_df = spark.read.parquet(ontology_metadata_parquet_path)
    gene_production_status_df = spark.read.parquet(gene_production_status_path)
    genotype_phenotype_df = spark.read.parquet(genotype_phenotype_parquet_path)
    impc_images_df = spark.read.parquet(impc_images_parquet_path)
    product_df = spark.read.parquet(product_parquet_path)
    gene_df: DataFrame = get_gene_core_df(
        imits_gene_df,
        imits_allele_df,
        mgi_homologene_df,
        mgi_mrk_list_df,
        embryo_data_df,
        observations_df,
        stats_results_df,
        ontology_metadata_df,
        gene_production_status_df,
        False,
    )
    gene_df = gene_df.withColumnRenamed(
        &#34;datasets_raw_data&#34;, &#34;gene_statistical_results&#34;
    )
    gene_df = gene_df.withColumn(
        &#34;_class&#34;, lit(&#34;org.mousephenotype.api.models.GeneBundle&#34;)
    )
    genotype_phenotype_df = genotype_phenotype_df.withColumnRenamed(
        &#34;marker_accession_id&#34;, &#34;mgi_accession_id&#34;
    )
    gp_by_gene_df = genotype_phenotype_df.groupBy(&#34;mgi_accession_id&#34;).agg(
        collect_set(
            struct(
                *[
                    col_name
                    for col_name in genotype_phenotype_df.columns
                    if col_name != &#34;mgi_accession_id&#34;
                ]
            )
        ).alias(&#34;gene_phenotype_associations&#34;)
    )
    gene_df = gene_df.join(gp_by_gene_df, &#34;mgi_accession_id&#34;, &#34;left_outer&#34;)

    impc_images_df = impc_images_df.withColumnRenamed(
        &#34;gene_accession_id&#34;, &#34;mgi_accession_id&#34;
    )

    images_by_gene_df = impc_images_df.groupBy(&#34;mgi_accession_id&#34;).agg(
        collect_set(
            struct(
                *[
                    col_name
                    for col_name in impc_images_df.columns
                    if col_name != &#34;mgi_accession_id&#34;
                ]
            )
        ).alias(&#34;gene_images&#34;)
    )
    gene_df = gene_df.join(images_by_gene_df, &#34;mgi_accession_id&#34;, &#34;left_outer&#34;)

    products_by_gene = product_df.groupBy(&#34;mgi_accession_id&#34;).agg(
        collect_set(
            struct(
                *[
                    col_name
                    for col_name in product_df.columns
                    if col_name
                    not in [&#34;mgi_accession_id&#34;] + EXCLUDE_PRODUCT_COLUMNS
                ]
            )
        ).alias(&#34;gene_products&#34;)
    )

    gene_df = gene_df.join(products_by_gene, &#34;mgi_accession_id&#34;, &#34;left_outer&#34;)
    gene_df = gene_df.withColumn(&#34;_id&#34;, col(&#34;mgi_accession_id&#34;))
    gene_df.write.format(&#34;mongo&#34;).mode(&#34;append&#34;).option(
        &#34;spark.mongodb.output.uri&#34;,
        f&#34;{self.mongodb_connection_uri}/admin?replicaSet={self.mongodb_replica_set}&#34;,
    ).option(&#34;database&#34;, str(self.mongodb_database)).option(
        &#34;collection&#34;, str(self.mongodb_genes_collection)
    ).save()

    ## Create gene search index
    gene_search_df = gene_df.select(
        col(&#34;mgi_accession_id&#34;).alias(&#34;_id&#34;),
        &#34;mgi_accession_id&#34;,
        &#34;marker_name&#34;,
        &#34;human_gene_symbol&#34;,
        &#34;marker_synonym&#34;,
        &#34;assignment_status&#34;,
        &#34;crispr_allele_production_status&#34;,
        &#34;es_cell_production_status&#34;,
        &#34;mouse_production_status&#34;,
        &#34;phenotype_status&#34;,
        &#34;phenotyping_data_available&#34;,
        col(&#34;significant_top_level_mp_terms&#34;).alias(&#34;significant_phenotype_system&#34;),
        col(&#34;not_significant_top_level_mp_terms&#34;).alias(
            &#34;non_significant_phenotype_system&#34;
        ),
        concat(
            &#34;gene_phenotype_associations.mp_term_id&#34;,
            flatten(&#34;gene_phenotype_associations.intermediate_mp_term_id&#34;),
            flatten(&#34;gene_phenotype_associations.top_level_mp_term_id&#34;),
        ).alias(&#34;significant_mp_term_ids&#34;),
        concat(
            &#34;gene_phenotype_associations.mp_term_name&#34;,
            flatten(&#34;gene_phenotype_associations.intermediate_mp_term_name&#34;),
            flatten(&#34;gene_phenotype_associations.top_level_mp_term_name&#34;),
        ).alias(&#34;significant_mp_term_names&#34;),
    )
    gene_search_df = gene_search_df.withColumn(
        &#34;_class&#34;, lit(&#34;org.mousephenotype.api.models.Gene&#34;)
    )
    gene_search_df.write.format(&#34;mongo&#34;).mode(&#34;append&#34;).option(
        &#34;spark.mongodb.output.uri&#34;,
        f&#34;{self.mongodb_connection_uri}/admin?replicaSet={self.mongodb_replica_set}&#34;,
    ).option(&#34;database&#34;, str(self.mongodb_database)).option(
        &#34;collection&#34;, str(self.mongodb_genes_collection) + &#34;_search&#34;
    ).save()
    gene_df.write.parquet(output_path)</code></pre>
</details>
</dd>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.output"><code class="name flex">
<span>def <span class="ident">output</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>The output that this Task produces.</p>
<p>The output of the Task determines if the Task needs to be run&ndash;the task
is considered finished iff the outputs all exist. Subclasses should
override this method to return a single :py:class:<code>Target</code> or a list of
:py:class:<code>Target</code> instances.</p>
<p>Implementation note
If running multiple workers, the output must be a resource that is accessible
by all workers, such as a DFS or database. Otherwise, workers might compute
the same output since they don't see the work done by other workers.</p>
<p>See :ref:<code>Task.output</code></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def output(self):
    return ImpcConfig().get_target(f&#34;{self.output_path}gene_bundle_parquet&#34;)</code></pre>
</details>
</dd>
<dt id="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.requires"><code class="name flex">
<span>def <span class="ident">requires</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>The Tasks that this Task depends on.</p>
<p>A Task will only run if all of the Tasks that it requires are completed.
If your Task does not require any other Tasks, then you don't need to
override this method. Otherwise, a subclass can override this method
to return a single Task, a list of Task instances, or a dict whose
values are Task instances.</p>
<p>See :ref:<code>Task.requires</code></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def requires(self):
    return [
        GeneExtractor(),
        AlleleExtractor(),
        MGIHomoloGeneExtractor(),
        MGIMrkListExtractor(),
        ObservationsMapper(),
        StatsResultsCoreLoader(raw_data_in_output=&#34;bundled&#34;),
        OntologyMetadataExtractor(),
        GeneProductionStatusExtractor(),
        GenotypePhenotypeCoreLoader(),
        ImpcImagesCoreLoader(),
        ProductExtractor(),
    ]</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<header>
<div style="max-width: 300px, text-align: center">
<img src="https://www.mousephenotype.org/wp-content/themes/impc/images/IMPC_10_YEAR_Logo.svg" alt="IMPC Logo">
</div>
</header>
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="impc_etl.jobs.load.impc_api" href="index.html">impc_etl.jobs.load.impc_api</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper">ImpcGeneBundleMapper</a></code></h4>
<ul class="">
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.app_options" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.app_options">app_options</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.embryo_data_json_path" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.embryo_data_json_path">embryo_data_json_path</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.main" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.main">main</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_connection_uri" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_connection_uri">mongodb_connection_uri</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_database" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_database">mongodb_database</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_genes_collection" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_genes_collection">mongodb_genes_collection</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_replica_set" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.mongodb_replica_set">mongodb_replica_set</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.name" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.name">name</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.output" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.output">output</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.output_path" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.output_path">output_path</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.packages" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.packages">packages</a></code></li>
<li><code><a title="impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.requires" href="#impc_etl.jobs.load.impc_api.impc_gene_bundle_mapper.ImpcGeneBundleMapper.requires">requires</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p><span></span></p>
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>