<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>impc_etl.jobs.extract.static_data_extractor API documentation</title>
<meta name="description" content="Static Data extractor
extract_human_gene_orthologues:
extract_phenotyping_centres:
extract_ontology_terms:" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>impc_etl.jobs.extract.static_data_extractor</code></h1>
</header>
<section id="section-intro">
<p>Static Data extractor
extract_human_gene_orthologues:
extract_phenotyping_centres:
extract_ontology_terms:</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">&#34;&#34;&#34;
Static Data extractor
    extract_human_gene_orthologues:
    extract_phenotyping_centres:
    extract_ontology_terms:
&#34;&#34;&#34;
import os
from typing import List
from pyspark.sql import DataFrame, SparkSession
from pyspark.sql.types import StructType, StructField, StringType, ArrayType
from owlready2 import (
    get_ontology,
    Ontology,
    onto_path,
    ThingClass,
    Nothing,
    Thing,
    IRIS,
)
from impc_etl.shared import utils
from impc_etl.shared.utils import convert_to_row
from impc_etl.config.ontology_schema import OntologySchema


def extract_human_gene_orthologues(
    spark_session: SparkSession, file_path: str
) -&gt; DataFrame:
    &#34;&#34;&#34;

    :param spark_session:
    :param file_path:
    :return human_gene_orthologues_df: Dataframe with the human gene to mouse gene mapping
    &#34;&#34;&#34;
    file_string_fields = [
        &#34;Human Marker Symbol&#34;,
        &#34;Human Entrez Gene ID&#34;,
        &#34;HomoloGene ID&#34;,
        &#34;Mouse Marker Symbol&#34;,
        &#34;MGI Marker Accession ID&#34;,
    ]
    file_array_fields = [&#34;High-level Mammalian Phenotype ID&#34;]
    schema_fields = [
        StructField(field_name, StringType(), True) for field_name in file_string_fields
    ]
    schema_fields.extend(
        [
            StructField(field_name, ArrayType(StringType), True)
            for field_name in file_array_fields
        ]
    )
    hmd_file_schema = StructType(schema_fields)
    human_gene_orthologues_df = utils.extract_tsv(
        spark_session, file_path, hmd_file_schema
    )
    return human_gene_orthologues_df


def extract_phenotyping_centres(
    spark_session: SparkSession, file_path: str
) -&gt; DataFrame:
    &#34;&#34;&#34;
    :param spark_session:
    :param file_path:
    :return:
    &#34;&#34;&#34;
    phenotyping_centres_df = utils.extract_tsv(spark_session, file_path)
    return phenotyping_centres_df


def extract_ontology_terms(
    spark_session: SparkSession, ontologies_path: str
) -&gt; DataFrame:
    &#34;&#34;&#34;

    :param spark_session:
    :param ontologies_path:
    :return:
    &#34;&#34;&#34;
    directory = os.fsencode(ontologies_path)
    ontology_terms = []
    for file in os.listdir(directory):
        filename = os.fsdecode(file)
        if filename.endswith(&#34;.owl&#34;):
            onto_path.append(os.path.join(directory, filename))
            ontology = get_ontology(None).load()
            ontology_terms.extend(parse_ontology(ontology))
    ontology_terms_df = spark_session.createDataFrame(
        convert_to_row(term) for term in ontology_terms
    )
    return ontology_terms_df


def parse_ontology(ontology: Ontology, schema=OntologySchema) -&gt; List[dict]:
    &#34;&#34;&#34;
    Parse an ontology from owlready2.Ontology to a list of dicts with
    the domain fields forOntologyTerm
    By default it uses the OBO Schema for the definition and synonyms annotations.
    :param ontology: owlready2.Ontology to parse
    :param schema: schema class extending OntologySchema
    :return ontology_terms: list of dicts containing ontology terms
    &#34;&#34;&#34;
    ontology_terms = []
    for ontology_class in ontology.classes():
        ontology_id = ontology.name
        ontology_term_id = ontology_class.name
        term_label = ontology_class.label
        term_definition = _collect_annotations(
            ontology_class, [schema.DEFINITION_ANNOTATION]
        )
        synonyms = _collect_annotations(ontology_class, schema.SYNONYM_ANNOTATIONS)
        parents = [
            str(parent.name)
            for parent in ontology_class.is_a
            if isinstance(parent, ThingClass)
        ]
        children = [
            str(child.name)
            for child in ontology_class.subclasses()
            if isinstance(child, ThingClass)
        ]
        ontology_term = {
            &#34;ontologyId&#34;: ontology_id,
            &#34;ontologyTermId&#34;: ontology_term_id,
            &#34;label&#34;: term_label,
            &#34;description&#34;: term_definition,
            &#34;synonyms&#34;: synonyms,
            &#34;parents&#34;: parents,
            &#34;children&#34;: children,
        }
        ontology_terms.append(ontology_term)
    return ontology_terms


def _collect_annotations(ontology_class: ThingClass, annotation_iris: List[str]):
    &#34;&#34;&#34;
    Collects the values for one or several annotations for one specific class
    :param ontology_class: owlready2.ThingClass
    :param annotation_iris: list of annotation iris
    :return annotations_values: list of values for the input annotations
    &#34;&#34;&#34;
    annotation_values = []
    for annotation_iri in annotation_iris:
        if IRIS[annotation_iri] is None or ontology_class in (Nothing, Thing):
            continue
        annotation_values.extend(IRIS[annotation_iri][ontology_class])
    return annotation_values</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="impc_etl.jobs.extract.static_data_extractor.extract_human_gene_orthologues"><code class="name flex">
<span>def <span class="ident">extract_human_gene_orthologues</span></span>(<span>spark_session: pyspark.sql.session.SparkSession, file_path: str) ‑> pyspark.sql.dataframe.DataFrame</span>
</code></dt>
<dd>
<div class="desc"><p>:param spark_session:
:param file_path:
:return human_gene_orthologues_df: Dataframe with the human gene to mouse gene mapping</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def extract_human_gene_orthologues(
    spark_session: SparkSession, file_path: str
) -&gt; DataFrame:
    &#34;&#34;&#34;

    :param spark_session:
    :param file_path:
    :return human_gene_orthologues_df: Dataframe with the human gene to mouse gene mapping
    &#34;&#34;&#34;
    file_string_fields = [
        &#34;Human Marker Symbol&#34;,
        &#34;Human Entrez Gene ID&#34;,
        &#34;HomoloGene ID&#34;,
        &#34;Mouse Marker Symbol&#34;,
        &#34;MGI Marker Accession ID&#34;,
    ]
    file_array_fields = [&#34;High-level Mammalian Phenotype ID&#34;]
    schema_fields = [
        StructField(field_name, StringType(), True) for field_name in file_string_fields
    ]
    schema_fields.extend(
        [
            StructField(field_name, ArrayType(StringType), True)
            for field_name in file_array_fields
        ]
    )
    hmd_file_schema = StructType(schema_fields)
    human_gene_orthologues_df = utils.extract_tsv(
        spark_session, file_path, hmd_file_schema
    )
    return human_gene_orthologues_df</code></pre>
</details>
</dd>
<dt id="impc_etl.jobs.extract.static_data_extractor.extract_ontology_terms"><code class="name flex">
<span>def <span class="ident">extract_ontology_terms</span></span>(<span>spark_session: pyspark.sql.session.SparkSession, ontologies_path: str) ‑> pyspark.sql.dataframe.DataFrame</span>
</code></dt>
<dd>
<div class="desc"><p>:param spark_session:
:param ontologies_path:
:return:</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def extract_ontology_terms(
    spark_session: SparkSession, ontologies_path: str
) -&gt; DataFrame:
    &#34;&#34;&#34;

    :param spark_session:
    :param ontologies_path:
    :return:
    &#34;&#34;&#34;
    directory = os.fsencode(ontologies_path)
    ontology_terms = []
    for file in os.listdir(directory):
        filename = os.fsdecode(file)
        if filename.endswith(&#34;.owl&#34;):
            onto_path.append(os.path.join(directory, filename))
            ontology = get_ontology(None).load()
            ontology_terms.extend(parse_ontology(ontology))
    ontology_terms_df = spark_session.createDataFrame(
        convert_to_row(term) for term in ontology_terms
    )
    return ontology_terms_df</code></pre>
</details>
</dd>
<dt id="impc_etl.jobs.extract.static_data_extractor.extract_phenotyping_centres"><code class="name flex">
<span>def <span class="ident">extract_phenotyping_centres</span></span>(<span>spark_session: pyspark.sql.session.SparkSession, file_path: str) ‑> pyspark.sql.dataframe.DataFrame</span>
</code></dt>
<dd>
<div class="desc"><p>:param spark_session:
:param file_path:
:return:</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def extract_phenotyping_centres(
    spark_session: SparkSession, file_path: str
) -&gt; DataFrame:
    &#34;&#34;&#34;
    :param spark_session:
    :param file_path:
    :return:
    &#34;&#34;&#34;
    phenotyping_centres_df = utils.extract_tsv(spark_session, file_path)
    return phenotyping_centres_df</code></pre>
</details>
</dd>
<dt id="impc_etl.jobs.extract.static_data_extractor.parse_ontology"><code class="name flex">
<span>def <span class="ident">parse_ontology</span></span>(<span>ontology: owlready2.namespace.Ontology, schema=impc_etl.config.ontology_schema.OntologySchema) ‑> List[dict]</span>
</code></dt>
<dd>
<div class="desc"><p>Parse an ontology from owlready2.Ontology to a list of dicts with
the domain fields forOntologyTerm
By default it uses the OBO Schema for the definition and synonyms annotations.
:param ontology: owlready2.Ontology to parse
:param schema: schema class extending OntologySchema
:return ontology_terms: list of dicts containing ontology terms</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def parse_ontology(ontology: Ontology, schema=OntologySchema) -&gt; List[dict]:
    &#34;&#34;&#34;
    Parse an ontology from owlready2.Ontology to a list of dicts with
    the domain fields forOntologyTerm
    By default it uses the OBO Schema for the definition and synonyms annotations.
    :param ontology: owlready2.Ontology to parse
    :param schema: schema class extending OntologySchema
    :return ontology_terms: list of dicts containing ontology terms
    &#34;&#34;&#34;
    ontology_terms = []
    for ontology_class in ontology.classes():
        ontology_id = ontology.name
        ontology_term_id = ontology_class.name
        term_label = ontology_class.label
        term_definition = _collect_annotations(
            ontology_class, [schema.DEFINITION_ANNOTATION]
        )
        synonyms = _collect_annotations(ontology_class, schema.SYNONYM_ANNOTATIONS)
        parents = [
            str(parent.name)
            for parent in ontology_class.is_a
            if isinstance(parent, ThingClass)
        ]
        children = [
            str(child.name)
            for child in ontology_class.subclasses()
            if isinstance(child, ThingClass)
        ]
        ontology_term = {
            &#34;ontologyId&#34;: ontology_id,
            &#34;ontologyTermId&#34;: ontology_term_id,
            &#34;label&#34;: term_label,
            &#34;description&#34;: term_definition,
            &#34;synonyms&#34;: synonyms,
            &#34;parents&#34;: parents,
            &#34;children&#34;: children,
        }
        ontology_terms.append(ontology_term)
    return ontology_terms</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<header>
<div style="max-width: 300px; text-align: center">
<img src="https://www.mousephenotype.org/wp-content/themes/impc/images/IMPC_10_YEAR_Logo.svg" alt="IMPC Logo">
</div>
<h1 style="text-align: center; max-width: 300px;">IMPC ETL</h1>
<h2 style="text-align: center; max-width: 300px;">Reference Documentation</h2>
</header>
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="impc_etl.jobs.extract" href="index.html">impc_etl.jobs.extract</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="impc_etl.jobs.extract.static_data_extractor.extract_human_gene_orthologues" href="#impc_etl.jobs.extract.static_data_extractor.extract_human_gene_orthologues">extract_human_gene_orthologues</a></code></li>
<li><code><a title="impc_etl.jobs.extract.static_data_extractor.extract_ontology_terms" href="#impc_etl.jobs.extract.static_data_extractor.extract_ontology_terms">extract_ontology_terms</a></code></li>
<li><code><a title="impc_etl.jobs.extract.static_data_extractor.extract_phenotyping_centres" href="#impc_etl.jobs.extract.static_data_extractor.extract_phenotyping_centres">extract_phenotyping_centres</a></code></li>
<li><code><a title="impc_etl.jobs.extract.static_data_extractor.parse_ontology" href="#impc_etl.jobs.extract.static_data_extractor.parse_ontology">parse_ontology</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p><span></span></p>
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>